src/continuous_agg.c

```diff
@@ -259,59 +358,165 @@ ts_continuous_agg_get_all_caggs_info(int32 raw_hypertable_id)
 		all_caggs_info.max_bucket_widths =
 			lappend(all_caggs_info.max_bucket_widths, DatumGetPointer(bucket_width));
 
+		bucket_function = palloc(sizeof(*bucket_function));
+		*bucket_function = cagg->bucket_function;
+		all_caggs_info.bucket_functions = lappend(all_caggs_info.bucket_functions, bucket_function);
+
 		all_caggs_info.mat_hypertable_ids =
 			lappend_int(all_caggs_info.mat_hypertable_ids, cagg->data.mat_hypertable_id);
 	}
 	return all_caggs_info;
 }
 
+/*
+ * Serializes ContinuousAggsBucketFunction* into a palloc'ated string like:
+ *
+ *     ver;bucket_width;origin;timezone
+ *
+ * ... where ver is a version of the serialization format. This particular format
+ * was chosen because of it's simplicity and good performance. Future versions
+ * of the procedure can use other format (like key=value or JSON) and/or add
+ * extra fields. NULL pointer is serialized to an empty string.
+ *
+ * Note that the schema and the name of the function are not serialized. This
+ * is intentional since this information is currently not used for anything.
+ * We can serialize the name of the function as well when and if this would be
+ * necessary.
+ */
+static const char *
+bucket_function_serialize(const ContinuousAggsBucketFunction *bf)
+{
+	const int ver = 1;
+	StringInfo str;
+
+	if (bf == NULL)
+		return pstrdup("");
+
+	str = makeStringInfo();
+
+	/* We are pretty sure that user can't place ';' character in these fields */
+	Assert(strstr(bf->bucket_width, ";") == NULL);
+	Assert(strstr(bf->origin, ";") == NULL);
+	Assert(strstr(bf->timezone, ";") == NULL);
+
+	appendStringInfo(str, "%d;%s;%s;%s;", ver, bf->bucket_width, bf->origin, bf->timezone);
+
+	return str->data;
+}
+
+/*
+ * Deserielizes a string into a palloc'ated ContinuousAggsBucketFunction*. Note
+ * that NULL is also a valid return value.
+ *
+ * See bucket_function_serialize() for more details.
+ */
+static ContinuousAggsBucketFunction *
+bucket_function_deserialize(const char *str)
+{
+	int i;
+	char *begin, *end, *strings[4];
+	ContinuousAggsBucketFunction *bf;
+
+	/* empty string stands for serialized NULL */
+	if (*str == '\0')
+		return NULL;
+
+	begin = pstrdup(str);
+	for (i = 0; i < lengthof(strings); i++)
+	{
+		end = strstr(begin, ";");
+		if (end == NULL)
+		{
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("failed to deserialize \"%s\" into a bucketing function", str),
+					 errdetail("separator not found")));
+		}
+
+		*end = '\0';
+		strings[i] = begin;
+		begin = end + 1;
+	}
+
+	/* end of string was reached */
+	Assert(*begin == '\0');
+
+	if (atoi(strings[0]) != 1)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("failed to deserialize \"%s\" into a bucketing function", str),
+				 errdetail("unsupported format version")));
+	}
+
+	bf = palloc(sizeof(ContinuousAggsBucketFunction));
+	bf->experimental = true;
+	bf->name = pstrdup("time_bucket_ng");
```

Comment:
Does it have to be `pstrdup()` here?

Reply:
I have mixed feelings about storing static value in a `char *` field. But I will check `const ContinuousAggsBucketFunction` can be returned here.

Reply:
The CPU cycle difference is significant here. It's either a pointer assignment or a full allocator call plus memory copy. Also, if this is called 100 times then it has 100x the memory footprint, whereas the static string is only allocated once. Plus memory allocator internal and external fragmentation overheads.

I can think of 2 reasons it would need to be in heap memory. 1 is if it needs to be modified and 2 is if it is explicitly `pfree()`ed somewhere.

Reply:
So I think it would be better to keep pstrdup() here for now. I added the comment that clarifies why it's used here.

Reply:
@mfundul Personally, I don't think we should invest too much time into optimizing variable-sized buckets now especially since no one reported this particular bottleneck yet. There could be hundreds of places that affect the performance even worse (network, disks, ... to name a few). Our current goal is to receive feedback from the early adopters. Also, I suspect that when the feature stabilizes we may end up without this field at all.

But if the team feels otherwise, I'm OK with it. Just let me know if you have strong objections here and I will replace it with a pointer to a static char[].

Reply:
I have no strong objections. But it's not really investing too much time in optimizing anything. It would be a trivial code simplification:
```
bf->name = pstrdup("time_bucket_ng");
```
vs
```
bf->name = "time_bucket_ng";
```

Reply:
OK, you convinced me. I also changed the return type of the function to `const ContinuousAggsBucketFunction` just to be on the safe side. 👍